### ISSUE TITLE
Make sure geocoder UI sends correct queries to nokia/here API

### ISSUE THREAD

**Issue by rafatower** [CartoDB/cartodb#1462]
Numbers are pretty bad for high-res geocoding in terms of geocoded rows. Make sure stats and queries are ok or fix them if they are not.

**Comment by juanignaciosl** [comment]
We're not using [API](https://developer.here.com/rest-apis/documentation/geocoder/topics/resource-geocode.html) fields like `country`, `state`/`county`, `street` but free form `searchtext`. This change is not trivial, since table column <-> field information is not even sent from the frontend right now, so change goes from the form to the geocoder. @iriberri , have you got some insights or examples about rows not geocoded or wrongly geocoded? I've tried the previous one (710 rideau road calgary, canada) with different fields and it returns the same point than with searchtext. cc @javisantana

**Comment by javisantana** [comment]
I see two solutions here:
- detect the country in the query, that's kind of easy if we only take into account english
- add a combo for country in the UI cc @saleiva

**Comment by iriberri** [comment]
I think there was a difference between 710 rideau road calgary, canada and 710 rideau road, calgary, canada.

What we also need to make sure is that if the user has three columns: `street`, `city`, `country`, we build the free text as: `{street}, {city}, {country}` always that the user select these columns in the different selectors.

What was happening with this specific Calgary user was that the comma between street and city wasn't added. This reduces a lot the success of Nokia in some examples I've done.

It's not easy to reproduce, it's in my to-do list to check why the commas weren't added in this example or in other similar issues we've had.

**Comment by juanignaciosl** [comment]
@javisantana UI already distinguish between those fields, the problem is it doesn't send it to the backend. I mean, it doesn't say "street column: street, country column: country" but "columns: street country", and since fields are not required you can't know which fields are. Changing how front sends the field would fix it, but you'd have to rewrite part of the backend as well (including current cache).

![screen shot 2015-01-23 at 11 40 01](https://cloud.githubusercontent.com/assets/932968/5873291/0d08088c-a2f5-11e4-886e-032e61cace11.png)

@iriberri well, if "710 rideau road calgary" was the `street` field api wouldn't even return a value (I can pass you the query if you want). The good side of using searchtext is that it's much tolerant to bad input.

**Comment by saleiva** [comment]
Yep, we did the actual UI in purpose in order to send the fields in the order we want.

**Comment by iriberri** [comment]
@juanignaciosl, in the table of the user, 710 rideau road and calgary were in different columns: this means that if we used the selector correctly, in the street we would only have 710 rideau road.

Take into account also that forcing the selectors to a parameter (country/province/street/whatever) would need to imply a small change: Right now, you cannot add a new extra selector if you haven't filled all the selectors.

This means that if the user wants to add a selector for a zipcode column but he/she doesn't have the province column filled, the process could be a little bit tricky in terms of UX :smile:

**Comment by saleiva** [comment]
Even being able to select extra fields before selecting the first ones, I don't believe people will get it. Also I don't think it fixes anything, since they would be in random order (the one that the user decides to select the fields on) and not in the order that we want, right?

**Comment by juanignaciosl** [comment]
@iriberri ok, that's what I thought the first time. I've tried api both with free form searchtext and with specific, different fields, and it returns the same point ([which seems correct now, this is the returned point](https://www.google.es/maps/place/505+30+Ave+SW,+Calgary,+AB+T2S+0P4,+Canada/@51.0261948,-114.073941,15z/data=!4m2!3m1!1s0x5371703dc5ec4cfd:0xcb48759d3f91a177)). Maybe they have better data now.

Buf, I'm using Nokia API and it's either too fuzzy or too strict. I was trying to use separate fields but I run into several problems:
* Country field is an exact iso match. We should have to do a name <-> iso transformation ourselves, and it should be fuzzy to allow bad user info, of course. I've tried to use `searchtext` for the country but, for example, `street=plaza+mayor&searchtext=spain` returns no result. Interestingly enough, `street=plaza+mayor&searchtext=mexico` does.
* `street=plaza+mayor&country=ESP` returns no result but `street=plaza+mayor&city=valladolid&country=ESP` does. A similar search with searchtext,  `searchtext=plaza+mayor%2C+españa` also works.

All that said, now I see the point of using `searchtext` instead of individual fields: it's more suitable for digesting user input. Switching to individual fields would lead to total number of geocoded rows. I think we're using it the best way we can. Switching to individual fields wouldn't return *more* hits at all, and it's not clear that they would be *better*. Unless you have more example of wrong or missed encodings I could try, I think we can't do anything better with Nokia API.

**Comment by juanignaciosl** [comment]
@saleiva : hadn't seen your answer: I wouldn't add fields I was speaking about not concatenating them, which is what we're doing now. We're not saying "street: s, country: c" but just "s, c", so we can't be sure at backend about what actually contains each column. Anyway, as I said, after having tried the API, I'm not sure this would fix anything at all.

**Comment by iriberri** [comment]
@juanignaciosl, we have a country decoder function for geocoding in data-services. In fact, we already use it in some other geocoding options as administrative regions or zipcodes.

I would focus on just checking that the commas issue, which I guess is a key fact in order to make Nokia able to separate the parts in their system. Probably they detect what are the pieces between the commas by themselves, so we could avoid doing it. :smile: It could be interesting to contact them to know the internals of both options.

Regarding the plaza mayor example: I suppose they have some kind of statistical function to return a value or not. If they have a huge amount of "plazas mayores" in Spain, maybe they just don't return any of them. By adding Valladolid to the text you're restricting the outcomes.

**Comment by juanignaciosl** [comment]
@iriberri commas comment was a good point, but I've tried adding and removing commas and it makes no difference, result is always right and the same, so I think it just works (both CartoDB and Nokia API, probably the problem back then was his data).

The point of Plaza Mayor was a counterexample of suitability of individual fields regarding "low number of results". Since it looks like searchtext field works in a "best effort" fashion (it will always return something if anything matches) and individual fields work more focused on exact matches using fields would only make our problem bigger.

**Comment by iriberri** [comment]
Could it be that they cache and process the results so when you try them a couple of times versions with commas and without are well recognized? 

I have a full table of Calgary streets where the results were always points in the same place: the center of Calgary city. What I did was to merge all the columns in a single one by using commas, and sent this column in a unique selector. The accuracy improved a lot -- I can provide both datasets if needed.

BTW, they don't always provide a result afaik. I've seen lots of null responses with valid addresses that couldn't be recognized by them.

**Comment by iriberri** [comment]
By the way, I just found this:

http://team.cartodb.com/u/iriberri/viz/d844d676-90e7-11e4-a662-0e0c41326911/embed_map

Green points are my attempt with commas. The blue (dark, because of frequency) dot is the initial attempt with several columns.

**Comment by iriberri** [comment]
I just created a specific issue to research on the comma's bug: https://github.com/CartoDB/cartodb/issues/1935

@Xatpy  :-)

**Comment by juanignaciosl** [comment]
I'll answer about comma issues there, then.

Well, it *could* be a cached result, but response doesn't say anything about it so we can only guess (or retry later). I don't think it's relevant anyway.

I know they don't always return a value, I said *if anything matches* ;-).

Insights of experimenting with the tables you sent.
`SELECT count(1), the_geom, array_agg(job_address) FROM iriberri.customer_list_282014_12_01_29_for_mapping_1 group by the_geom having count(1) > 10`
* There are 2 points repeated: one (-114.08529,51.05011) has 242 points, other has 142. The second one is due to missing addresses, so I'll speak about the first one: it *looks* like certainly there're problems with those addresses.
* Experiments with that addresses (I've used a different one after each experiment to avoid your cache guess):
  * Using the api manually:
    * No commas at all: 51.0541191, -114.0380783
    * Comma between city and country: 51.033989, -114.1351013
    * Comma between address, city and country: 51.0120392, -114.0726776
  * Using CartoDB:
    * (Our) Cache hits:
      * Specifying country column: 51.1208611, -114.1623688
      * Specifying country by hand: 51.075779, -114.0761108
    * Avoiding our cache by adding 1 to the street number
      * Specifying country column: 51.1654091, -114.2649155 (query is sent with commas between street, city and country)
      * Specifying country by hand: 51.0708694, -114.0463715 (query is also sent with commas between those three columns)

Since it looks like "it works now", I've exported your table, cleaned the_geom and cartodb_georef_status and geocoded: all rows where geocoded, and the only matching points (same query than before) are those with empty streets.

Summing up, I think those 7 cases cover every possible situation, and all requests returned different results, and my geocoding has worked, so I'm pretty sure this works ok.

What might have happened?
* The user chose the wrong columns. Did you attempt to geolocate it straight away, to check whether user just made a mistake?
* Nokia API has improved since then.
* CartoDB has improved since then.
* "Temporal glitch" (sorry) of Nokia or CartoDB.

1.- Do you agree this is ok and it's working then, or am I missing something?
2.- Assuming *yes* to 1, what do you think might have happened?

**Comment by iriberri** [comment]
This is getting messy, let me answer inline:

>The user chose the wrong columns. Did you attempt to geolocate it straight away, to check whether user just made a mistake?

We have the exact logs for this. I even tried this in the own user account by myself when this happened.  No mistakes I'd say. :-)
> Nokia API has improved since then.

Blue points and green points were geocoded the same day. The thing is that I've also tried to replicate this "bug" and I haven't been able: in the logs the commas appear each time I've tried out of that user's account.
> CartoDB has improved since then.

That part hasn't been touched for months.
>"Temporal glitch" (sorry) of Nokia or CartoDB.

My guess is just getting to know why when I did the geocodings in the user account and in my account, in the exact same way, our logs are different.

User's account:
`ESTE_USER | high-resolution | 415 | {job_address} {job_city}, canada`

My account:
`iriberri | high-resolution | 6 | {job_address}, {job_city}, canada`

> Using the api manually: No commas at all: 51.0541191, -114.0380783

We agree that the commas are important for Nokia, then. They weren't added in that case: Could this be provoked by some kind of race condition or whatever in our end? :confused:

**Comment by juanignaciosl** [comment]
Ok to your test, nokia not improved and us not improved.

I think [the comment](https://github.com/CartoDB/cartodb/issues/1935#issuecomment-71197276) at #1935 closes the mistery: if you use that "additional information to complete the *address*" field that information is not separated by a comma. I think changing that text to "additional information to complete the *street* address".
![screen shot 2015-01-23 at 15 05 28](https://cloud.githubusercontent.com/assets/932968/5875852/d71d4df0-a311-11e4-87b9-af4c1497674e.png)

So @iriberri was totally right about comma importance, but I think we're doing it right. I'll wait for your opinion to close both. And thank you very much for your help :-)

**Comment by iriberri** [comment]
Awesome finding, @juanignaciosl. Although this is making me crazy: I don't remember doing that, but I'm sure this solves the mistery about why I haven't been able to reproduce it later by using the UI adequately.

We need to add the comma too for that selector: it is thought for zipcodes, for example, which also need to be separated. 

Again, awesome. :heart_eyes:  I've been puzzled about this for weeks.

P.S.: Just as an extra proof of the missing commas: take into account that we could even know what text is being returned by checking in our internal cache: commas weren't added in the same way that them didn't appear in the formatter. The screenshot in the first post is from the cache.

**Comment by juanignaciosl** [comment]
Your wish is my command, @iriberri :) I've just deployed the fix to staging (rui) so you can try (please let me know when you've finished).

Anyway I'll also ask @xavijam for the CR, since he's the author of that geocoding parametrisation and we might not have everything in mind.

**Comment by saleiva** [comment]
Hey @juanignaciosl could we make sure we have metrics for this, to see if it's really improving the geocoding results? Is that already measured in kibana? We could create a widget just for this

**Comment by juanignaciosl** [comment]
I've just deployed the change that fixes #1465 , that will give us insights on automatic triggered geocodings. I'll improve [current dashboard](http://ela01.useb.cartodb.net/#/dashboard/elasticsearch/Geocoding) in a few days so we can see automatic results vs non-automatic. There's also a `formatter` column with this information so you can get more information about the usage of this.
